### PR TITLE
C1 mode B added, AES CTR decryption, CRC calculation for C1 mode B 

### PR DESCRIPTION
**Pull request by Jacman777**

### CODE REVIEW

File: mbus_packet.cpp [hunk 1]
```diff
@@ -261,33 +261,22 @@ uint16_t verifyCrcBytesCmodeB(uint8_t* pByte, uint8_t* pPacket, uint16_t packetS
...
   uint16_t i = 0;
-  if (packetSize > 128) {
-    while (i < 126) {
-      crc = crcCalc(crc, pByte[i]);
-      pPacket[i] = pByte[i];
-      ++i;
-    }
-
-    if ((~crc) != (pByte[i] << 8 | pByte[i + 1])) {
+  //Support only for packets with L (lenght byte) < 128 (packetSize < 131)
```
Comment (SzczepanLeon) [inline mbus_packet.cpp:264]: why?

File: mbus_packet.cpp [hunk 2]
```diff
@@ -261,33 +261,22 @@ uint16_t verifyCrcBytesCmodeB(uint8_t* pByte, uint8_t* pPacket, uint16_t packetS
...
-    }
-
-    pPacket[i] = pByte[i];
-    ++i;
-    pPacket[i] = pByte[i];
-    ++i;
-    crc = 0;
   }
-
+  // To compensate incrrect packet length eg. L=33 -> L=31
```
Comment (SzczepanLeon) [inline mbus_packet.cpp:268]: Why it is incorrect?

Comment (Jacman777) [reply mbus_packet.cpp:268]: Good question. Probably incorrect is not right word here. It's just different to C1 A frame. 
Following below docu, L-field in C1 B frame includes 2 bytes of CRC checksum. Which is not included in C1 A frame. 
image
https://www.st.com/resource/en/application_note/an4772-wmbus-2013-firmware-stack-overview-stmicroelectronics.pdf

File: mbus_packet.cpp (hunk 1 above)
Comment (Jacman777) [reply mbus_packet.cpp:264]: I wanted to consider here only C1 B frames which consists of first and second block only. (Ignoring frames with additional optional blocks).
image

**Comment by SzczepanLeon** [comment]
Probably I will refactor my code soon. Can You send me some samples with C frames type B?

**Comment by Jacman777** [comment]
> Probably I will refactor my code soon. Can You send me some samples with C frames type B?

Sure, here you go (including logs from ESPHOME):
```
08:34:24.777 ->  FullFrame: 0x54, 0x3D, 0x33, 0x44, 0x2D, 0x2C, 0x06, 0x50, 0x21, 0x62, 0x19, 0x04, 0x8D, 0x20, 0x12, 0x50, 0xD6, 0x2D, 0x23, 0x8F, 0x12, 0x97, 0x6D, 0xFC, 0x40, 0xA3, 0x1E, 0x48, 0xF5, 0xED, 0xE9, 0x8F, 0x25, 0x71, 0x90, 0x45, 0x5A, 0xC4, 0x5C, 0xF7, 0xA5, 0xC9, 0x09, 0x26, 0x63, 0x20, 0x15, 0x45, 0x2D, 0x1B, 0x5C, 0xF7, 
08:34:24	[I]	[wmbus:092]	
Using driver 'kamheat' for ID [0x62215006] RSSI: -67 dBm LQI: 135 Mode: C1 T: 31442D2C0650216219048D201250D62D238F12976DFC40A31E48F5EDE98F257190455AC45CF7A5C90926632015452D1B5CF7 (50)

08:37:41.039 ->  FullFrame: 0x54, 0x3D, 0x33, 0x44, 0x2D, 0x2C, 0x06, 0x50, 0x21, 0x62, 0x19, 0x04, 0x8D, 0x20, 0x14, 0xA8, 0xD6, 0x2D, 0x23, 0x0A, 0xBF, 0xE4, 0x14, 0x54, 0xFE, 0x4E, 0x95, 0x9F, 0x44, 0x50, 0x40, 0x6B, 0x3A, 0xC9, 0x56, 0x60, 0x4C, 0x59, 0x7F, 0x9D, 0x85, 0x05, 0x27, 0xCB, 0x18, 0x20, 0x59, 0xF1, 0xFB, 0x5D, 0x15, 0x8A, 
08:37:41	[I]	[wmbus:092]	
Using driver 'kamheat' for ID [0x62215006] RSSI: -71 dBm LQI: 136 Mode: C1 T: 31442D2C0650216219048D2014A8D62D230ABFE41454FE4E959F4450406B3AC956604C597F9D850527CB182059F1FB5D158A (50)

08:40:57.131 ->  FullFrame: 0x54, 0x3D, 0x33, 0x44, 0x2D, 0x2C, 0x06, 0x50, 0x21, 0x62, 0x19, 0x04, 0x8D, 0x20, 0x16, 0xD0, 0xD6, 0x2D, 0x23, 0xDD, 0x2D, 0x28, 0x8F, 0x5C, 0x65, 0x65, 0xB1, 0x1E, 0x7C, 0x30, 0x86, 0x0D, 0x19, 0xAF, 0x2B, 0x9E, 0x33, 0xE8, 0x64, 0x97, 0x2F, 0x68, 0xE9, 0x95, 0x5A, 0x5D, 0x0D, 0x96, 0x1F, 0x9B, 0xF2, 0x2D, 
08:40:57	[I]	[wmbus:092]	
Using driver 'kamheat' for ID [0x62215006] RSSI: -65 dBm LQI: 132 Mode: C1 T: 31442D2C0650216219048D2016D0D62D23DD2D288F5C6565B11E7C30860D19AF2B9E33E864972F68E9955A5D0D961F9BF22D (50)

08:42:35.091 ->  FullFrame: 0x54, 0x3D, 0x33, 0x44, 0x2D, 0x2C, 0x06, 0x50, 0x21, 0x62, 0x19, 0x04, 0x8D, 0x20, 0x17, 0xF4, 0xD6, 0x2D, 0x23, 0xF8, 0x8F, 0xD8, 0xD4, 0x15, 0x85, 0xC1, 0xA4, 0x6B, 0x99, 0x05, 0x47, 0x60, 0x07, 0x6D, 0xA7, 0xD6, 0xE8, 0x4E, 0x32, 0xEB, 0x58, 0x01, 0xC5, 0x0F, 0xF8, 0x84, 0x4F, 0xBB, 0x40, 0xC4, 0x5C, 0x15, 
08:42:35	[I]	[wmbus:092]	
Using driver 'kamheat' for ID [0x62215006] RSSI: -68 dBm LQI: 136 Mode: C1 T: 31442D2C0650216219048D2017F4D62D23F88FD8D41585C1A46B99054760076DA7D6E84E32EB5801C50FF8844FBB40C45C15 (50)

08:44:13.028 ->  FullFrame: 0x54, 0x3D, 0x33, 0x44, 0x2D, 0x2C, 0x06, 0x50, 0x21, 0x62, 0x19, 0x04, 0x8D, 0x20, 0x18, 0x08, 0xD7, 0x2D, 0x23, 0x26, 0xAD, 0xE6, 0x80, 0x46, 0x8B, 0x14, 0x65, 0x2D, 0xDB, 0x9C, 0x11, 0x54, 0xF4, 0xB9, 0x1D, 0x3B, 0xD4, 0x21, 0x1A, 0xB4, 0x89, 0xCD, 0x8D, 0xBB, 0x08, 0x06, 0x61, 0xDA, 0xBC, 0x44, 0xAD, 0xC3, 
08:44:13	[I]	[wmbus:092]	
Using driver 'kamheat' for ID [0x62215006] RSSI: -66 dBm LQI: 137 Mode: C1 T: 31442D2C0650216219048D201808D72D2326ADE680468B14652DDB9C1154F4B91D3BD4211AB489CD8DBB080661DABC44ADC3 (50)

08:49:06.510 ->  FullFrame: 0x54, 0x3D, 0x42, 0x44, 0x2D, 0x2C, 0x06, 0x50, 0x21, 0x62, 0x19, 0x04, 0x8D, 0x20, 0x1B, 0x54, 0xD7, 0x2D, 0x23, 0x84, 0xD8, 0x3F, 0xFE, 0x3B, 0xCD, 0xFB, 0x8F, 0x87, 0x78, 0x82, 0x15, 0x84, 0x81, 0x1D, 0x7E, 0x15, 0xFA, 0x07, 0xC7, 0xD6, 0xB9, 0x6E, 0x29, 0xD8, 0x56, 0x56, 0x21, 0x2E, 0xC6, 0x78, 0x7C, 0x83, 0x19, 0xC9, 0xEC, 0xA9, 0x97, 0xE6, 0x20, 0x1C, 0x76, 0x06, 0x10, 0xB0, 0xFD, 0xDB, 0x90, 
08:49:06	[I]	[wmbus:092]	
Using driver 'kamheat' for ID [0x62215006] RSSI: -64 dBm LQI: 129 Mode: C1 T: 40442D2C0650216219048D201B54D72D2384D83FFE3BCDFB8F8778821584811D7E15FA07C7D6B96E29D85656212EC6787C8319C9ECA997E6201C760610B0FDDB90 (65)
```

**Comment by SzczepanLeon** [comment]
Will be delivered in wmbus component version 3, together with refactoring.

**Pull request closed without merging**